File: src/game_modes/deps/mem.c
#include <stdint.h>
#include "mem.h"
#include "drawing_api.h"
#include "minimal_state_info.h"

extern char GAME_MODE_NAME[256];
extern uint8_t NUM_TEAMS;
extern uint8_t NUM_DRAWABLE_OBJECTS;
MinimalTeamInfo TEAMS[255];
MinimalMapInfo MAP;

MinimalPlayerInfo PLAYER_TO_BE_ADDED;
MinimalTeamInfo WINNING_TEAM;
DrawableObject DRAWABLE_OBJECTS[256];

uint64_t RNG_SEED;

uint64_t drawable_objects_ptr(void) {
	return (uint64_t)DRAWABLE_OBJECTS;

}

uint64_t num_drawable_objects(void) {
	return NUM_DRAWABLE_OBJECTS;

}

uint64_t player_to_be_added_ptr(void) {
	return (uint64_t)&PLAYER_TO_BE_ADDED;

}

uint64_t winning_team_ptr(void) {
	return (uint64_t)&WINNING_TEAM;

}

uint64_t teams_ptr(void) {
	return (uint64_t)TEAMS;

}

uint32_t num_teams(void) {
	return NUM_TEAMS;

}

uint64_t name_ptr(void) {
	return (uint64_t)GAME_MODE_NAME;

}

uint64_t map_ptr(void) {
	return (uint64_t)&MAP;

}


void set_rng_seed(uint64_t seed) {
	RNG_SEED = seed;

}

uint64_t rand(void) {
	return(((RNG_SEED = RNG_SEED * 214013L + 2531011L) >> 16) & 0x7fff);

}

uint64_t rand_range_u64(uint64_t min, uint64_t max) {
        return min + (rand() % (max - min));

}

void* memset(void* s, int c, unsigned long n) {
	char* sb = s;

	for (char* z = s; z < sb + n; z += 1) {
		*(char*)z = c;

	}

	return s;

}

void* memcpy(void* dst, const void *restrict src, unsigned long n) {
	const char *restrict sb = src;
	char *db = dst;

	for (unsigned long i = 0; i < n; i += 1) {
		*(db + i) = *(sb + i);

	}
	
	return dst;
		
}

File: src/game_modes/capture_the_flag/src/mem.rs
include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
use crate::NUM_TEAMS;
use nanorand::rand::wyrand::WyRand;

const ZERO_DRAWABLE_OBJECT: DrawableObject = DrawableObject {
    pos_x: 0.0,
    pos_y: 0.0,
    width: 0.0,
    height: 0.0,
    color: 0,
    shape: 0,

};

const ZERO_PLAYER: MinimalPlayerInfo = MinimalPlayerInfo {
    id: 0,
    team_id: 0,
    num_kills: 0,
    using_ability: 0,
    health: 0,
    ammo: 0,
    ability: 0,
    cloaking: false,
    direction: 0.0,
    pos_x: 0.0,
    pos_y: 0.0,
    weapon: 0,
    throwing_grenade: false as u16,
};

const ZERO_TEAM: MinimalTeamInfo = MinimalTeamInfo {
    id: 0,
    score: 0,
    players: [ZERO_PLAYER; 255],
    num_players: 0,
    color: 0,
};

const ZERO_MAP_OBJ: MinimalMapObject = MinimalMapObject {
    pos_x: 0.0,
    pos_y: 0.0,
    width: 0.0,
    height: 0.0,

};

const ZERO_MAP: MinimalMapInfo = MinimalMapInfo {
    width: 0.0,
    height: 0.0,
    num_spawn_points: 0,
    spawn_points: [ZERO_MAP_OBJ; 256],

};

pub static mut TEAMS: [MinimalTeamInfo; 2] = [ZERO_TEAM; 2];
pub static mut RNG: WyRand = WyRand::new_seed(0);
pub static mut MAP: MinimalMapInfo = ZERO_MAP;
pub static mut DRAWABLE_OBJECTS: [DrawableObject; 255] = [ZERO_DRAWABLE_OBJECT; 255];
pub static mut PLAYER_TO_BE_ADDED: MinimalPlayerInfo = ZERO_PLAYER;
pub static mut WINNING_TEAM: MinimalTeamInfo = ZERO_TEAM;

#[no_mangle]
pub unsafe fn drawable_objects_ptr() -> u64 {
    DRAWABLE_OBJECTS.as_mut_ptr() as u64

}

#[no_mangle]
pub unsafe fn num_drawable_objects() -> u64 {
    2

}

#[no_mangle]
pub unsafe fn player_to_be_added_ptr() -> u64 {
    (&mut PLAYER_TO_BE_ADDED) as *mut MinimalPlayerInfo as u64

}

#[no_mangle]
pub unsafe fn winning_team_ptr() -> u64 {
    (&mut WINNING_TEAM) as *mut MinimalTeamInfo as u64

}

#[no_mangle]
pub unsafe fn teams_ptr() -> u64 {
    TEAMS.as_mut_ptr() as u64

}

#[no_mangle]
pub unsafe fn num_teams() -> u64 {
    NUM_TEAMS as u64

}

#[no_mangle]
pub unsafe fn map_ptr() -> u64 {
    (&mut MAP) as *mut MinimalMapInfo as u64

}

#[no_mangle]
pub unsafe fn set_rng_seed(seed: u64) {
    RNG = WyRand::new_seed(seed);

}

static mut GAME_MODE_NAME: [u8; 17] = [0; 17];

#[no_mangle]
pub unsafe fn name_ptr() -> u64 {
    GAME_MODE_NAME[0..16].as_mut().copy_from_slice(crate::GAME_MODE_NAME.as_bytes());
    GAME_MODE_NAME.as_ptr() as u64

}


